parquet-column/src/main/java/org/apache/parquet/schema/Types.java

```diff
@@ -555,114 +622,1267 @@ protected GroupType build(String name) {
+        parent.setKeyType(build("key"));
+        return new MapMapValueBuilder<P, Q>(parent).repetition(repetition);
+      }
+
+      public MapMapValueBuilder<P, Q> requiredMapValue() {
+        return mapValue(Type.Repetition.REQUIRED);
+      }
+
+      public MapMapValueBuilder<P, Q> optionalMapValue() {
+        return mapValue(Type.Repetition.OPTIONAL);
+      }
+
+      public MapListValueBuilder<P, Q> listValue(Type.Repetition repetition) {
+        parent.setKeyType(build("key"));
+        return new MapListValueBuilder<P, Q>(parent).repetition(repetition);
+      }
+
+      public MapListValueBuilder<P, Q> requiredListValue() {
+        return listValue(Type.Repetition.REQUIRED);
+      }
+
+      public MapListValueBuilder<P, Q> optionalListValue() {
+        return listValue(Type.Repetition.OPTIONAL);
+      }
+
+      public BaseMapBuilder<P, Q> value (Type type) {
+        parent.setValueType(type);
+        return this.parent;
+      }
+
+      public BaseMapBuilder<P, Q> requiredValue (Type type) {
+        return value(type);
+      }
+
+      public BaseMapBuilder<P, Q> optionalValue (Type type) {
+        return value(type);
+      }
+    }
+
+    public static class MapGroupValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseGroupBuilder<P, MapGroupValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapGroupValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapMapValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseMapBuilder<P, MapMapValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapMapValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapMapValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapListValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseListBuilder<P, MapListValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapListValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapListValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapListKeyBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
```

Comment:
If this requires exposing the `keyElement` methods on the list builder to avoid problems with `named`, then lets not support lists as map keys in the API. We can add it back later, but the implementation it quite complicated and I think it is cleaner to allow callers to simply set a list type instead of having so much code. I think this should allow you to remove all of the key/element code and this will be much shorter.

We can always add this back in later.